editor/plugins/org.fusesource.ide.camel.editor/src/org/fusesource/ide/camel/editor/features/create/ext/CreateFigureFeature.java

```diff
@@ -118,35 +128,63 @@ public CATEGORY_TYPE getCategoryType() {
 	public boolean canCreate(ICreateContext context) {
 		ContainerShape container = context.getTargetContainer();
 		Object containerBO = getBusinessObjectForPictogramElement(container);
+		
+		// creating a figure on the Camel Context / Diagram directly (for instance Route)
 		if (container instanceof Diagram) {
-			// we want to create something on the diagram rather than inside a container element - thats only allowed on camelContexts
 			CamelFile cf = ((CamelDesignEditor)getDiagramBehavior().getDiagramContainer()).getModel();
-			if (cf.getCamelContext() == null && cf.getChildElements().size()==0) {
+			// sanity check if the CamelFile has a Camel Context
+			if (cf.isEmpty()) {
+				// if not, then we need to add one right away to prevent further problems
 				cf.addChildElement(new CamelContextElement(cf, null));
 			}
-			if (eip != null) return eip.getName().equalsIgnoreCase("route") || 
-				eip.getName().equalsIgnoreCase("rest") || eip.getName().equalsIgnoreCase("restConfiguration");
-			if (clazz != null) { 
+			
+			// if we got an EIP reference then we can check if thats allowed
+			// to be added to the camel context
+			if (eip != null) {
+				return eip.canBeAddedToCamelContextDirectly();
+			// if we only have a class defined
+			} else if (clazz != null) { 
+				// then we need to instantiate it
 				Object obj = newInstance(clazz);
-				if (obj != null && obj instanceof AbstractCamelModelElement) return ((AbstractCamelModelElement)obj).getNodeTypeId().equalsIgnoreCase("route");
+				if (obj != null && 
```

Comment:
obj instanceof AbstractCamelModelElement is enough, no need to check also if != null